### ISSUE TITLE
Cyclic type class bad error message

### ISSUE THREAD

**Issue by LiamGoodacre** [purescript/purescript#3223]
The following code:
```purescript
class A <= B
class B <= A
```

Generates the following error:
```
A cycle appears in a set of type synonym definitions.
Cycles are disallowed because they can lead to loops in the type checker.
Consider using a 'newtype' instead.
```

This confusingly talks about synonyms and newtype, but not type classes.

**Comment by cdparks** [comment]
It looks to me like this happens because the compiler collects binding groups after desugaring typeclasses to type synonyms. Should checking for cycles in superclasses be a separate step before desugaring?

**Comment by hdgarrood** [comment]
This was resolved by #3381.
